### PR TITLE
Unify ArtistDetailsRestController and SearchArtistRestController

### PR DESCRIPTION
**Pull request by hedgehog1833**
Removed:
- SearchArtistRestController
- ArtistDetailsRestController
- FollowArtistRestController
- FollowArtistService

New:
- ArtistsRestController

Extended:
- ArtistsService

https://trello.com/c/e2ika2TU

### CODE REVIEW

**Comment by hedgehog1833** [comment]
[Unify ArtistDetailsRestController and SearchArtistRestController](https://trello.com/c/e2ika2TU/51-unify-artistdetailsrestcontroller-and-searchartistrestcontroller)

File: src/main/java/com/metalr2/service/artist/ArtistsServiceImpl.java [hunk 1]
```diff
@@ -33,23 +58,117 @@ public ArtistsServiceImpl(ArtistsRepository artistsRepository) {
...
+  public boolean followArtist(long discogsId) {
+    Optional<DiscogsArtist> artistOptional = artistSearchClient.searchById(discogsId);
+
+    if (artistOptional.isEmpty()) {
+      return false;
+    }
+
+    FollowedArtistEntity followedArtistEntity       = new FollowedArtistEntity(currentUserSupplier.get().getPublicId(),
+                                                                              artistOptional.get().getName(), discogsId);
+    FollowedArtistEntity savedFollowedArtistEntity  = followedArtistsRepository.save(followedArtistEntity);
```
Comment (DanielW1987) [inline src/main/java/com/metalr2/service/artist/ArtistsServiceImpl.java]: `savedFollowedArtistEntity` is unused.

File: src/main/java/com/metalr2/service/artist/ArtistsServiceImpl.java [hunk 2]
```diff
@@ -33,23 +58,117 @@ public ArtistsServiceImpl(ArtistsRepository artistsRepository) {
...
+    Optional<FollowedArtistEntity> optionalFollowedArtistEntity = followedArtistsRepository.findByPublicUserIdAndArtistDiscogsId(
+        currentUserSupplier.get().getPublicId(), discogsId);
+
+    if (optionalFollowedArtistEntity.isEmpty()){
+      return false;
+    }
+
+    FollowedArtistEntity followedArtistEntity = optionalFollowedArtistEntity.get();
+    followedArtistsRepository.delete(followedArtistEntity);
+    return true;
```
Comment (DanielW1987) [inline src/main/java/com/metalr2/service/artist/ArtistsServiceImpl.java]: Java 8 functional style
```
optionalFollowedArtistEntity.ifPresent(followedArtistsRepository::delete);
return optionalFollowedArtistEntity.isPresent();
```

File: src/main/java/com/metalr2/service/artist/ArtistsServiceImpl.java [hunk 3]
```diff
@@ -33,23 +58,117 @@ public ArtistsServiceImpl(ArtistsRepository artistsRepository) {
...
+  @Override
+  public Optional<ArtistNameSearchResponse> searchDiscogsByName(String artistQueryString, int page, int size) {
+    Optional<DiscogsArtistSearchResultContainer> responseOptional = artistSearchClient.searchByName(artistQueryString, page, size);
+
+    if (responseOptional.isEmpty()) {
+      return Optional.empty();
+    }
+
+    ArtistNameSearchResponse response = mapNameSearchResult(responseOptional.get());
+    return Optional.of(response);
```
Comment (DanielW1987) [inline src/main/java/com/metalr2/service/artist/ArtistsServiceImpl.java]: You can use the map() method of Optional here.

File: src/main/java/com/metalr2/service/artist/ArtistsServiceImpl.java [hunk 4]
```diff
@@ -33,23 +58,117 @@ public ArtistsServiceImpl(ArtistsRepository artistsRepository) {
...
+  @Override
+  public Optional<ArtistDetailsResponse> searchDiscogsById(long artistId) {
+    Optional<DiscogsArtist> responseOptional = artistSearchClient.searchById(artistId);
+
+    if (responseOptional.isEmpty()) {
+      return Optional.empty();
+    }
+
+    ArtistDetailsResponse response = mapDetailsSearchResult(responseOptional.get());
+    return Optional.of(response);
```
Comment (DanielW1987) [inline src/main/java/com/metalr2/service/artist/ArtistsServiceImpl.java]: You can use the map() method of Optional here.

File: src/main/java/com/metalr2/web/controller/rest/ArtistsRestController.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+    return success ? ResponseEntity.status(HttpStatus.CREATED).build() : ResponseEntity.notFound().build();
+  }
+
+  @DeleteMapping(path = Endpoints.Rest.UNFOLLOW + "/{discogsId}")
+  public ResponseEntity<Void> handleUnfollow(@PathVariable long discogsId) {
+    boolean success = artistsService.unfollowArtist(discogsId);
+
+    return success ? ResponseEntity.ok().build() : ResponseEntity.notFound().build();
+  }
```
Comment (DanielW1987) [inline src/main/java/com/metalr2/web/controller/rest/ArtistsRestController.java]: It doesn't feel right yet that we push the following and unfollowing of artists into the REST paradigm in this way. In the end, it is always an action that is carried out on an artist.

I would suggest that both following and unfollowing be executed via POST. If successful, there are 200, if not found 404.

File: src/main/java/com/metalr2/service/artist/ArtistsService.java [hunk 6]
```diff
@@ -1,16 +1,25 @@
...
 
-  Optional<ArtistDto> findByArtistDiscogsId(long artistDiscogsId);
+  Optional<ArtistDto> findArtistByDiscogsId(long discogsId);
+  List<ArtistDto> findAllArtistsByDiscogsIds(long... discogsIds);
+  boolean existsArtistByDiscogsId(long discogsId);
 
-  List<ArtistDto> findAllByArtistDiscogsIdIn(long... artistDiscogsIds);
+  boolean followArtist(long discogsId);
+  boolean unfollowArtist(long discogsId);
+  boolean isFollowed(long discogsId);
```
[the remark this answers is not in the document]
Comment (hedgehog1833) [reply src/main/java/com/metalr2/service/artist/ArtistsService.java]: No it's not. I changed it to discogsId

File: src/main/java/com/metalr2/service/artist/ArtistsService.java [hunk 7]
```diff
@@ -1,16 +1,25 @@
...
+  List<FollowArtistDto> findFollowedArtistsPerUser(String publicUserId);
 
-  boolean existsByArtistDiscogsId(long artistDiscogsId);
+  Optional<ArtistNameSearchResponse> searchDiscogsByName(String artistQueryString, int page, int size);
+  Optional<ArtistDetailsResponse> searchDiscogsById(long artistId);
```
[the remark this answers is not in the document]
Comment (hedgehog1833) [reply src/main/java/com/metalr2/service/artist/ArtistsService.java]: Hmm sure about that? I think it is good to know that these methods search an external source and not our own database.

File: src/main/java/com/metalr2/web/controller/rest/ArtistsRestController.java (hunk 5 above)
Comment (hedgehog1833) [reply src/main/java/com/metalr2/web/controller/rest/ArtistsRestController.java]: Sounds about right to me, I'll change it!

File: src/main/java/com/metalr2/service/artist/ArtistsService.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (DanielW1987) [reply src/main/java/com/metalr2/service/artist/ArtistsService.java]: In my view, this is a detail of our architecture that may change in the future. The method does not necessarily have to know this.

**Pull request merged**